pkg/oauthclientscontroller/oauthclientscontroller.go

```diff
@@ -0,0 +1,147 @@
+package oauthclientscontroller
+
+import (
+	"monis.app/go/openshift/controller"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	configv1 "github.com/openshift/api/config/v1"
+	oauthv1 "github.com/openshift/api/oauth/v1"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+	configv1client "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	configinformer "github.com/openshift/client-go/config/informers/externalversions/config/v1"
+	oauthclient "github.com/openshift/client-go/oauth/clientset/versioned/typed/oauth/v1"
+	oauthinformer "github.com/openshift/client-go/oauth/informers/externalversions/oauth/v1"
+	routeclient "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+	routeinformer "github.com/openshift/client-go/route/informers/externalversions/route/v1"
+
+	authopclient "github.com/openshift/cluster-authentication-operator/pkg/operatorclient"
+	authoputil "github.com/openshift/cluster-authentication-operator/pkg/util"
+)
+
+type oauthClientsController struct {
+	authOperatorClient authopclient.OperatorClient
+
+	authConfigClient configv1client.AuthenticationsGetter
+
+	oauthClientGetter oauthclient.OAuthClientsGetter
+	routeClient       routeclient.RouteInterface
+}
+
+type clientEnsureFunc func(*oauthClientsController, *oauthv1.OAuthClient, *routev1.Route) error
+
+func NewOAuthClientsController(
+	authOpConfigClient authopclient.OperatorClient,
+	authConfigClient configv1client.AuthenticationsGetter,
+	authConfigInformer configinformer.AuthenticationInformer,
+	oauthClientGetter oauthclient.OAuthClientsGetter,
+	oauthClientInformer oauthinformer.OAuthClientInformer,
+	routeClient routeclient.RouteInterface,
+	routeInformer routeinformer.RouteInformer,
+) controller.Runner {
+	c := &oauthClientsController{
+		authOperatorClient: authOpConfigClient,
+
+		authConfigClient: authConfigClient,
+
+		oauthClientGetter: oauthClientGetter,
+		routeClient:       routeClient,
+	}
+
+	targetNameFilter := controller.FilterByNames(nil, authoputil.TargetName)
+	configNameFilter := controller.FilterByNames(nil, authoputil.GlobalConfigName)
```

Comment:
And I do not think this makes sense at all because you cannot make a parent func for it.